### PR TITLE
Allow relative imports from URL-submitted CWL

### PR DESCRIPTION
**Pull request by cjllanwarne**
- [x] Needs a rebase once #4283 merges
- [x] Needs to bring in changes to heterodon to include the SSL modules for https:// URLs
- [ ] Switch the `cwl_format_valid_with_workflow_url.test` back to point at `develop` once this merges

### CODE REVIEW

File: languageFactories/cwl-v1-0/src/main/scala/languages/cwl/CwlV1_0LanguageFactory.scala [hunk 1]
```diff
@@ -28,15 +30,25 @@ class CwlV1_0LanguageFactory(override val config: Config) extends LanguageFactor
                                  workflowIdForLogging: WorkflowId,
                                  ioFunctions: IoFunctionSet,
                                  importResolvers: List[ImportResolver]): Parse[ValidatedWomNamespace] = {
+
+    def parse(): Parse[Cwl] = source.workflowUrl match {
+      case Some(url) => for {
+        reference <- fromEither[IO](CwlReference.fromString(url).map(Right(_)).getOrElse(Left(NonEmptyList.one(s"Invalid workflow reference: $url")))): Parse[CwlReference]
```
Comment (aednichols) [inline languageFactories/cwl-v1-0/src/main/scala/languages/cwl/CwlV1_0LanguageFactory.scala]: It seems like somewhere in all this abstraction we must be making an HTTP requests to actually get the `url`. How is it doing that?

File: docs/developers/CwlParsing.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# How CWL is Parsed
+
+The process of parsing CWL runs approximately like this:
+
+1. Produce a canonical one-file representation of the CWL as a JSON object in memory 
+1. Use Circe to turn that representation into scala case classes
```
Comment (kshakir) [inline docs/developers/CwlParsing.md]: ToL: When this is ready to merge, consider switching the numbers from `1. 1. 1. …` to `1. 2. 3. …`. There are a dozens of us who read md files using `less`.

File: docs/developers/CwlParsing.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    - Optional fields in case classes are allowed to not exist in the JSON
+    - All fields in the JSON must be represented by fields in the appropriate case classes
+    - In cases where the JSON may be structured in one of several ways, we use Shapeless coproducts to specify "one of" many options.
+
+If a file is failing to parse the first stop should always be to double-check that the case classes accurately represent
+the range of possible CWL JSON.
+
+## Convert the case classes into WOM representations
+
+As of today, CWL doesn't produce WOM bundles. Instead, it goes straight to executable.
```
Comment (kshakir) [inline docs/developers/CwlParsing.md]: ToL: Not 100% sure what `it goes straight to executable` means. If that makes sense in to CWL developers then LGTM.

File: cwl/src/main/scala/cwl/CwltoolRunner.scala [hunk 4]
```diff
@@ -26,13 +26,13 @@ object CwltoolRunner {
...
-  override def salad(file: BFile): String = {
-    val commandResult: CommandResult = %%("cwltool", "--quiet", "--print-pre", file.toString)
+  override def salad(reference: CwlReference): String = {
+    val commandResult: CommandResult = %%("cwltool", "--quiet", "--print-pre", reference.pathAsString)
     commandResult.exitCode match {
       case 0 => commandResult.out.string
       case error =>
         throw new RuntimeException(
-          s"running CwlTool on file $file resulted in exit code $error and stderr ${commandResult.err.string}")
+          s"running CwlTool on file $reference resulted in exit code $error and stderr ${commandResult.err.string}")
```
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/CwltoolRunner.scala]: It does, and I've updated this.

Reference includes the "which tool/workflow in the file to look at" information, which is not relevant for a message about an unreadable file.

File: languageFactories/cwl-v1-0/src/main/scala/languages/cwl/CwlV1_0LanguageFactory.scala (hunk 1 above)
Comment (cjllanwarne) [reply languageFactories/cwl-v1-0/src/main/scala/languages/cwl/CwlV1_0LanguageFactory.scala]: Eventually all this goes into a call to `cwltool`, which accepts commands in the form of `cwltool --print-pre URL` and gives us back an appropriate CWL string

Comment (aednichols) [reply languageFactories/cwl-v1-0/src/main/scala/languages/cwl/CwlV1_0LanguageFactory.scala]: Ah OK, I was hunting around in our Scala code for this but it makes total sense that we're handing it off

**Pull request merged**